cosima_cookbook/explore.py

```diff
@@ -0,0 +1,965 @@
+                          self.widgets['daterange']],
+                          layout={'padding': '10% 0', 'width': '50%'})
+
+        centre_pane = HBox([VBox([self.widgets['var_selector']]),
+                                  info_pane])
+
+        # Call super init and pass widgets as children
+        super().__init__(children=[self.widgets['header'],
+                                   self.widgets['expt_selector'],
+                                   centre_pane,
+                                   self.widgets['load_button'],
+                                   self.widgets['data_box']])
+
+    def _set_handlers(self):
+        """
+        Define routines to handle button clicks and experiment selection
+        """
+
+        self.widgets['load_button'].on_click(self._load_data)
+        self.widgets['expt_selector'].observe(self._expt_eventhandler, names='value')
+
+    def _expt_eventhandler(self, selector):
+        """
+        Called when experiment dropdown menu changes
+        """
+        self._load_experiment(selector.new)
+
+    def _load_data(self, b):
+        """
+        Called when load_button clicked
+        """
+
+        data_box = self.widgets['data_box']
+
+        varname = self.widgets['var_selector'].get_selected()
+        (start_time, end_time) = self.widgets['daterange'].value
+        frequency = self.widgets['frequency'].value
+
+        load_command = """
+        <pre><code>cc.querying.getvar('{expt}', '{var}', session, 
+                    start_time='{start}', end_time='{end}', frequency='{frequency}')</code></pre>
+        """.format(expt=self.widgets['expt_selector'].value, 
+                var=varname,
+                start=str(start_time),
+                end=str(end_time),
+                frequency=str(frequency))
+
+        # Interim message to tell user what is happening
+        data_box.value = 'Loading data, using following command ...\n\n' + load_command + 'Please wait ... '
+
+        try:
+            self.data = querying.getvar(self.experiment_name,
+                                        varname,
+                                        self.de.session, 
+                                        start_time=str(start_time),
+                                        end_time=str(end_time),
+                                        frequency=frequency)
+        except Exception as e:
+            data_box.value = data_box.value + 'Error loading variable {} data: {}'.format(varname, e)
+            return
+
+        # Update data box with message about command used and pretty HTML
+        # representation of DataArray
+        data_box.value = 'Loaded data with' + load_command + self.data._repr_html_()
+
+    def _load_experiment(self, experiment_name):
+        """
+        When first instantiated, or experiment changed, the variable
+        selector widget needs to be refreshed
+        """
+        self.de = DatabaseExtension(self.session, experiments=experiment_name)
```

Comment:
Is this what you meant by having to re-cache the database contents when experiments are changed? Maybe `DatabaseExtension` could cache the full thing once, and provide an interface to narrow its view to certain experiments (instead of recreation?)

Reply:
The whole extension thing is a horrible hack. I was hoping to re-implement most of the guts of it in the ORM layer if I had time, which I evidently do not, as this has taken WAY too long.

When I'm testing I make a single `de` object and pass it into the widgets so never have to recalculate. I wasn't sure about always generating the whole thing every time, as it is possible to pass a list of experiments to it as well, and this is what happens when you just make an `ExperimentExplorer` and pass it an experiment name. So that it is quick, as it just pulls in the information for the one experiment. It is important that it only has the info for a single experiment, as we don't want it picking up variables from other experiments.

The `DatabaseExtension` thing is doing too much work. Like I said, hacky. For `DatabaseExplorer` it makes the mapping between `(name, long_name)` and `experiment` for filtering. But we don't need that for the `ExperimentExplorer`.

I should rethink that, but am happy for any input, as I've been staring at this stuff for too long and am a bit over it TBH.

I'm trying to write tests ATM, which should be useful if there is some refactoring done.

Reply:
Thanks for all the feedback too. Very helpful.